### PR TITLE
add information about git-based pkgs

### PR DESCRIPTION
**Pull request by jagajaga**

### CODE REVIEW

File: doc/coding-conventions.xml [hunk 1]
```diff
@@ -256,6 +256,12 @@ bound to the variable name <varname>e2fsprogs</varname> in
   a package named <literal>hello-svn</literal> by
   <command>nix-env</command>.</para></listitem>
 
+  <listitem><para>If package is fetched from git's commit then
+  the version part of the name <emphasis>must</emphasis> be the date of that 
+  (fetched) commit. The date must be in <literal>"YYYY-MM-DD"</literal> format.
+  Also add <literal>"git"</literal> to the name - e.g.,
+  <literal>"pkgname-git-2014-09-23"</literal>.</para></listitem>
```
Comment (peti) [inline doc/coding-conventions.xml]: Huh? Why not spell the date YYYY.MM.DD? That seems to me like it's much more in line with our usual interpretation of what a "version number" looks like.

Comment (jagajaga) [inline doc/coding-conventions.xml]: Good idea.
But I will have to rename lots of pkgs :anguished:
Anyway - OK, looks nice.

**Comment by jagajaga** [comment]
@peti fixed

**Comment by bjornfor** [comment]
This could quickly become a bikeshed fest, but ...

I think dashes are nice because we use this mechanism as a replacement for a version number, something to use when there actually is no version number to begin with. Some projecs actually do use YYYY.MM date-like version numbers. Maybe it will be less confusing if we make the date aspect of the number really explicit?

Dashed date format is an ISO standard: http://en.wikipedia.org/wiki/ISO_8601. Well, to be fair, no-dash is also in that standard, but not the 'dot' style.

**Comment by peti** [comment]
Hmm, the diff still says

> The date must be in <literal>"YYYY-MM-DD"</literal> format.

**Comment by jagajaga** [comment]
@peti @bjornfor oh, dashes are an ISO standard and I like them more too. So let's leave them :smile:

**Comment by peti** [comment]
Well, from a technical point of view, both spellings are equally good (as far as nix-env is concerned). It will interpret `2015.02.23` like any other version number. `2015-02-23` will be interpreted as `2014` being the version number and `02-23` being an additional tag attached to it. That's why I find the first option more aesthetically pleasing in the context of `nix-env`. Ultimately, both approaches give the same observable behavior, though ...

**Comment by bjornfor** [comment]
Right. And my idea of the date being visually distinct from a version number isn't that important if we add "-git" to the package name (as suggested). It should be pretty obvious that a "-git" package is not a proper release. I don't have strong feelings (I think).

**Comment by jagajaga** [comment]
@bjornfor usually it's nice to see the date. Because there can be problems in a program and if we see a date - we can check for updates.

**Comment by bjornfor** [comment]
Of course, I wasn't suggesting to remove the date.

**Comment by peti** [comment]
OK. I believe it's good to have a consistent scheme recommended in the guidelines, and both YYYY-MM-DD and YYYY.MM.DD seem sensible to me. Maybe the ISO8601 format is a tad nicer.  For what it's worth, I'll support either choice. I think @jagajaga should decide. Then we can blame him later when something goes wrong. :grinning:

**Comment by jagajaga** [comment]
Dashes! :+1:

**Comment by bjornfor** [comment]
@peti: +1.

One complication though: What to do when packaging a git version *in addition to the relased version*? (I'm not particularly fond of this, but some people seem to want git snapshots into nixpkgs tree. So it may be nice to have a standard to adhere to.)

I suggest that the corresponding git snapshot of ```pkgname-X.Y``` would be named ```pkgname-git-X.Y-COMMITS_SINCE_RELEASE```. Basically similar to what "git describe" gives you in a properly tagged tree. (I'm a bit uncertain about whether to use the "-git" suffix in this case, but that's mostly irrelevant.)

Reasoning: it is cumbersome to compare version and date (need to do manual lookup for when the version was released); so we better make it obvious how the two packages relate to each other.

**Comment by jagajaga** [comment]
BTW check this https://github.com/NixOS/nixpkgs/issues/6479

**Comment by jagajaga** [comment]
@bjornfor the ideas of `COMMITS_SINCE_RELEASE ` seems to be nice and easy to implement. I think we have to provide this things in `nix-prefetch-git`.

**Comment by edolstra** [comment]
If we're documenting current practice: the most prevalent form of date-based version in Nixpkgs at the moment is `YYYYMMDD` (i.e. no interpunction):
```
$ nix-env -qa | grep '201[0-9][0-9][0-9]' | wc -l
207

$ nix-env -qa | grep '201[0-9]-[0-9][0-9]' | wc -l
73

$ nix-env -qa | grep '201[0-9]\.[0-9][0-9]' | wc -l
22
```

**Comment by jagajaga** [comment]
I'm working on changing it.

**Comment by vcunat** [comment]
@edolstra: I didn't realize we have so many non-delimited dates in versions. However, I think it's better to document what we *prefer* there in future than to blindly cement the currently dominating style. For me personally it's hard to read.

IMHO:
- there shouldn't be too many packages where we provide **both** released and unreleased versions;
- I don't think it really matters where the source comes from, be it git, a prerelease tarball, etc.;
- A good version name slightly depends on upstream versioning conventions, but it's good to try to have meaningful linear ordering preserved, e.g. something like ``1.4pre-${DATE}``. Reference for the comparison algorithm: http://nixos.org/nix/manual/#ssec-version-comparisons; note that we already have a special status for "pre" components.

**Comment by jagajaga** [comment]
So what are we going to document? :)

**Comment by rushmorem** [comment]
I think we should document and enforce non-delimited dates. Here is why...

 - *Numbers don't lie*. As @edolstra pointed out, packages using non-delimited dates are more than double that of the other 2 candidates combined. This might be due to a single or a few collaborators' effort or it might be indicative of what the majority of the collaborators prefer, it doesn't matter. If we treat each of these packages as votes, the winner is very clear. We all have our personal preferences, but I think this will be the more objective decision we can make.
 - *Parser friendliness*. Granted, all three versions are undoubtedly easy to parse and work with. However, non-delimited dates are by far the easiest. Once you isolate them you can cast and virtually treat them as integers without any further effort! Less code, less opportunity for silly bugs...
 - *Collaborator friendliness*.  Yes, in isolation they are harder to read than the other 2 options but once you combine them with a package name (which can contain hyphens) and a file extension (which contains at least one period), I find them easier to spot (and read in that context). Combined with their parser friendliness, you can quickly convert them into any of the other 2 formats if you wish.

**Comment by rushmorem** [comment]
I think @vcunat also raised an important point that it should not matter where the source comes from. If I have the same version of my source code in git, svn, tarball etc, it's going to produce the same binary. I second his suggestion to use something generic for the prefix.

**Comment by jagajaga** [comment]
@rushmorem the idea of `git` prefix comes from the point that the program isn't `stable` as we can understand it. So `git` prefix tells us that this pkg is from bleeding edge sources. We can see the same situation in `AUR`. And dates tell us when was this pkg fetched from a source.
Why I'm so into dashes - because we can definetly say that it's a date and easily parse it by ourselves. _Parses friendliness_ - oh, let's write as easy as we can, just to be sure we won't have bugs! That's stupidity, IMO. The dashes and non-dashes parsers are gonna be epsilon different :smile:

**Comment by rushmorem** [comment]
@jagajaga I appreciate the need for dates and prefixes. I'm not suggesting that we don't use them. On the contrary I think it's a good idea to use them.

My point is that rather than using `git` as a prefix for sources using git as their version control, `svn` for subversion etc, why not use something generic descriptive like `${version}pre-${date}`. For example if the I wanted to snapshot the kernel today I would use something like `4.0pre-20140225`. If we wanted to make sure that this is just our own version not upstream's then we could use something like `nixpre` instead of just `pre`. The previous example will then become `4.0nixpre-20140225`.

The advantage with a generic prefix is that the source can change version control systems (or wherever they come from) without having to change our version in the nix packages tree but still keeping the semantic meaning of that version.

If we have this convention in place (which is what we are trying to achieve anyway) then we should always know for sure that what follows `nixpre-` is a date. Otherwise it should be treated as a bug in packaging.

**Comment by jagajaga** [comment]
@rushmorem what about pkg with no version at all? There are lot's of them.

**Comment by rushmorem** [comment]
@jagajaga same thing applies. In that case `${upcomingVersion}` is `null` so the prefix would just be `nixpre`. It would fit exactly the same bill like using just `git` as a prefix, however, with the nice benefit that if the package does have some form of versioning in place we quickly what version we are building towards without having to look it up.

**Comment by jagajaga** [comment]
Idea of changing `git` prefix to a `nixpre` seems to be good. But why `nix`pre?

**Comment by rushmorem** [comment]
@jagajaga good catch. In the case where we are not building towards any particular version then `pre` doesn't make sense at all. How about we use just `nix` then?

**Comment by jagajaga** [comment]
@rushmorem so why not use any random word? :)
I've chosen `git` because of `AUR` as I mention earlier. I think lots of users will recognise that.

**Comment by rushmorem** [comment]
I [used](https://github.com/NixOS/nixpkgs/issues/4290) Arch as my main distro for a long time before I discovered NixOS. The issue I have with `git` is that it's suggestive. It suggests that this particular version is a git snapshot. If you don't use it like that then you lose semantic meaning. The whole time I was using Arch, whenever, I saw a package ending with git I that's what I thought. If you look at the version column of their [AUR packages](https://aur.archlinux.org/packages/?O=0&C=0&SeB=n&K=git&outdated=&SB=n&SO=a&PP=50&do_Search=Go) you will see there is a lot of confusion there.

Right now it doesn't matter much because git is vastly popular and almost everyone is using it. What happens when, say in 5/10 years another version control system becomes more  popular than git? Subversion once was popular too. Yes we can continue to use it like a "random word" but it will be confusing to a lot of people. This is why I was suggesting a word that doesn't suggest a particular VCS. People would be thinking, "OMG! These guys are still using git. Who does that in this day and age!" :)

Another prefix we could use, for example, is `nixpkg`. We could use just that for both packages with and without versioning. What I like about `nix` or `nixpkg` prefixes is that like `git` they are suggestive. But they suggest that this is just a nixpkgs repo version, not an upstream version.

Anyway, this is just my opinion. The decision is ultimately yours, so you are free to go with what you prefer.

**Comment by nckx** [comment]
Some form of standard date-based versioning for the vast majority of VCS packages is sorely needed. I also agree with adding a mandatory, uniform tag, whether it's `-nixpkg` or something else, to all "unofficial" releases. 

*That said...*

**< :bike: >**

Some packages use `git describe` or similar to provide a more descriptive version string in their interface or documentation. In those specific cases, it would be a tiny step back to enforce a different scheme at the packaging level, and it would be nice to have a (documented!) way around that **must**.

**</ :bike: >**

**Comment by jagajaga** [comment]
Ping, guys 
Are we going to merge it this way or make changes? We have already add date output to nix-prefetch-git.

**Comment by domenkozar** [comment]
I'll try to review this during pycon sprints.

**Comment by jagajaga** [comment]
@domenkozar thank you.

**Comment by edwtjo** [comment]
FTR; I really dislike the dashes naming for the reasons that @peti/@edolstra declared.

**Comment by jagajaga** [comment]
@edwtjo well. I prefer dashes of dots. We need some kind of poll, because indeed there is no difference between these formats. But with no interpunction it's harder to read.

**Comment by jagajaga** [comment]
But dots - they are for real versions. Dates are usually consists of dashes.

**Comment by edwtjo** [comment]
@jagajaga Yes, I've read the thread, although apparently a couple of months to late. There is a difference between the formats since their evaluation is different.. I also agree on @rushmorem point about not using a SCM dependent prefix, if any at all should be used. Note for now that I'm promoting the dashes naming in PRs because; wu wei.

I agree that a poll is in order to settle these kind of issues.

**Comment by jagajaga** [comment]
Seems that mostly people follow these rules. Merging?

**Comment by vcunat** [comment]
I'm OK with it. The formulation really works only for master of git-controlled projects, but that seems to be the vast majority of use cases. We might figure out the rest later, when we encounter it more. 

The "pre"-style (see above) has some advantages IMO, but it really depends on how people typically want to use the package with nix-env, and there I don't see a strong preference, so perhaps it's useless to try to find some complex rules ATM.

*A nitpick: I would personally replace "then" by a comma (or just add the comma).*

**Comment by jagajaga** [comment]
Going to merge in a day, ok?

**Pull request merged**